keps/sig-auth/20190916-noderestriction-pods.md

```diff
@@ -0,0 +1,309 @@
+- Solving all possible man-in-the-middle attacks.
+
+## Proposal
+
+All restrictions will be enforced through the [NodeRestriction][] admission controller. These
+extensions will be guarded by the `NodeRestrictionPods` feature gate.
+
+On update, only the delta will be restricted. In other words, the Kubelet can modify pods with
+restricted labels / annotations / owner references as long as it's not modifying (or adding or
+deleting) the restricted entries.
+
+### Label Restrictions
+
+The Kubelet will be prevented from updating pod labels or creating mirror pods with labels, except
+for whitelisted keys:
+
+- Any labels starting with `unrestricted.node.kubernetes.io/` are allowed.
+
+The Kubelet does not currently label pods, nor are there official label keys that apply to
+pods. However, there are a few labels that are commonly applied to system addons & static pods:
+
+- `component` (used by [kubeadm][kubeadm-labels])
+- `tier` (used by [kubeadm][kubeadm-labels])
+- `k8s-app` (common on [addons][addons-k8s-app])
+
+The `k8s-app` label is used to match controllers for system components, and therefore should be
+explicitly disallowed.
+
+It is not clear how the `tier` and `component` labels are consumed, but I recommend uses be migrated
+to `unrestricted.node.kubernetes.io/component` and `unrestricted.node.kubernetes.io/tier`, rather
+than special casing the existing labels.
+
+[kubeadm-labels]: https://github.com/kubernetes/kubernetes/blob/e682310dcc5d805a408e0073e251d99b8fe5c06d/cmd/kubeadm/app/util/staticpod/utils.go#L60
+[addons-k8s-app]: https://github.com/kubernetes/kubernetes/blob/e682310dcc5d805a408e0073e251d99b8fe5c06d/cluster/addons/kube-proxy/kube-proxy-ds.yaml#L23
+
+### Annotation Restrictions
+
+The Kubelet will be prevented from updating pod annotations or creating mirror pods with
+annotations, except for whitelisted keys:
+
+1. Any annotations starting with `unrestricted.node.kubernetes.io/` are allowed.
+2. Annotations the Kubelet currently uses are allowed:
+- `ConfigMirrorAnnotationKey = "kubernetes.io/config.mirror"`
+- `ConfigHashAnnotationKey = "kubernetes.io/config.hash"`
+- `ConfigFirstSeenAnnotationKey = "kubernetes.io/config.seen"`
+- `ConfigSourceAnnotationKey = "kubernetes.io/config.source"`
+3. Well-known annotations that may be used on static pods are allowed:
+- `PodPresetOptOutAnnotationKey = "podpreset.admission.kubernetes.io/exclude"
+- `SeccompPodAnnotationKey = "seccomp.security.alpha.kubernetes.io/pod"`
+- `SeccompContainerAnnotationKeyPrefix = "container.seccomp.security.alpha.kubernetes.io/"` (prefix
+  match)
+- `ContainerAnnotationKeyPrefix = "container.apparmor.security.beta.kubernetes.io/"` (prefix match)
+- `PreferAvoidPodsAnnotationKey = "scheduler.alpha.kubernetes.io/preferAvoidPods"`
+- `BootstrapCheckpointAnnotationKey = "node.kubernetes.io/bootstrap-checkpoint"`
+
+Note that annotation restrictions are applying the least-privilege principal, but are not directly
+mitigating a known threat.
```

Comment:
The case for restricting annotations doesn't seem that strong to me, given the number of exempted annotations, and the lack of a known threat that is being mitigated.

Reply:
Done. I moved the existing content to alternatives considered, to be revisited if/when we come up with a more concrete use case.

Other options I considered are using a whitelist approach (dubious without use cases), and forbidding annotation changes in pod/status updates (limits node identity reuse without clear benefit).